File: VSCode/Codigo03_MODELOLINEAR.py
import tclab
import numpy as np
import time
import matplotlib.pyplot as plt

# Conexão com o TCLab
a = tclab.TCLab()

# Tempo de execução em minutos
run_time = 20

# Número de ciclos (1 ciclo por segundo)
loops = int(60.0 * run_time)
tm = np.zeros(loops)            # Vetor para armazenar o tempo
T1 = np.zeros(loops)            # Temperatura real
T1_ModeloNL = np.zeros(loops)   # Temperatura modelada não linear
T1_ModeloL = np.zeros(loops)    # Temperatura modelada linear
Q1 = np.ones(loops) * 50        # Potência fixa em 50%
u = np.zeros(loops)             # Entrada no modelo linear

a.Q1(50)    # Aplicar potência ao hardware
Q = 50      # Aplicar potência aos modelos

# Parâmetros do modelo não linear
Tambiente = 301.15  # 28 °C em Kelvin
Alpha = 0.01
Cp = 500
A = 0.0012
m = 0.004
U = 8
Emissividade = 0.9
Boltzmann = 5.67e-8

# Gráfico
plt.figure(figsize=(12, 8))
plt.ion()
plt.show()

# Loop principal
start_time = time.time()
prev_time = start_time
dt = 1.0  # intervalo de tempo em segundos

try:
    for i in range(loops):
        # Tempo de espera
        sleep_time = max(0.01, dt - (time.time() - prev_time))
        time.sleep(sleep_time)

        # Atualizar tempo
        t = time.time()
        prev_time = t
        tm[i] = t - start_time

        # Ler temperatura real do TCLab
        T1[i] = a.T1

        # --- Modelo Não Linear ---
        if i == 0:
            T1_ModeloNL[i] = T1[0]
        else:
            Taquecedor = T1_ModeloNL[i-1] + 273.15
            dTdt = ((Alpha / (m * Cp)) * Q) + ((U * A) / (m * Cp)) * (Tambiente - Taquecedor) \
                   + ((Emissividade * Boltzmann * A) / (m * Cp)) * (Tambiente**4 - Taquecedor**4)
            T1_ModeloNL[i] = T1_ModeloNL[i-1] + dTdt * dt

        # --- Modelo Linear ---
        if i == 0:
            T1_ModeloL[i] = T1[0]  # condição inicial
        else:
            T1_ModeloL[i] = T1_ModeloL[i-1] + 0.00499 * np.exp(-tm[i] / 123) * Q


        # Atualizar gráfico
        plt.clf()
        ax1 = plt.subplot(2, 1, 1)
        ax1.grid()
        plt.plot(tm[:i+1], T1[:i+1], 'r-', label='Temperatura Real', linewidth=2)
        plt.plot(tm[:i+1], T1_ModeloNL[:i+1], 'b--', label='Modelo Não Linear', linewidth=2)
        plt.plot(tm[:i+1], T1_ModeloL[:i+1], 'g--', label='Modelo Linear', linewidth=2)
        plt.ylabel('Temperatura (°C)', fontsize=14)
        plt.legend(loc='best', fontsize=12)

        ax2 = plt.subplot(2, 1, 2)
        ax2.grid()
        plt.plot(tm[:i+1], Q1[:i+1], 'r-', label='Potência (%)', linewidth=2)
        plt.ylabel('Potência (%)', fontsize=14)
        plt.xlabel('Tempo (s)', fontsize=14)
        plt.legend(loc='best', fontsize=12)

        plt.tight_layout()
        plt.draw()
        plt.pause(0.05)

    # --- Cálculo do Erro e Salvamento ---
    erro_medio_NL = np.mean(np.abs(T1 - T1_ModeloNL))
    erro_medio_L = np.mean(np.abs(T1 - T1_ModeloL))
    print(f'\nErro Médio Absoluto entre Temperatura Real e Modelo Não Linear: {erro_medio_NL:.2f} °C')
    print(f'Erro Médio Absoluto entre Temperatura Real e Modelo Linear: {erro_medio_L:.2f} °C')

    # Salvar gráfico em PNG
    plt.savefig('Grafico_TCLab_Modelo_Linear.png', dpi=300, bbox_inches='tight')

    # Salvar dados em .txt
    np.savetxt('Dados_simulacao_03.txt', np.column_stack((tm, T1, T1_ModeloNL, T1_ModeloL, Q1)),
               header='Tempo (s)\tTemperatura Real (°C)\tModelo Não Linear (°C)\tModelo Linear (°C)\tPotência (%)', fmt='%.2f', delimiter='\t')

except KeyboardInterrupt:
    print('Interrompido pelo usuário.')

except Exception as e:
    print(f'Erro: {e}')
    raise

finally:
    a.Q1(0)    # Desliga o aquecedor
    a.LED(0)   # Apaga LED
    a.close()
    print('Simulação concluída.')

File: MATLAB/Codigo03_modelo_linear.m
clear; clc;

%% Inicialização do hardware
run('tclab.m'); % Arquivo com funções T1C(), h1(), led()

% Garantir desligamento ao final, mesmo com Ctrl+C ou erro
cleanupObj = onCleanup(@() desligarTCLab());

%% Parâmetros de Simulação
run_time = 20;                 % Tempo total da simulação (min)
loops = 60 * run_time;         % Número de amostras (1 Hz)
dt = 1.0;                      % Intervalo de tempo (s)

% Inicialização de variáveis
tm = zeros(loops,1);           % Vetor de tempo (s)
T1 = zeros(loops,1);           % Temperatura real (°C)
T1_ModeloNL = zeros(loops,1);  % Temperatura modelada não linear (°C)
T1_ModeloL = zeros(loops,1);   % Temperatura modelada linear (°C)
Q1 = ones(loops,1) * 50;       % Potência aplicada (%)
u = zeros(loops,1);            % Entrada no modelo linear (caso necessário)

% Aplicar potência inicial ao hardware
h1(Q1(1));

% Potência aplicada aos modelos
Q = 50;

%% Parâmetros do Modelo Não Linear
Tambiente = 301.15;  % Temperatura ambiente (K) - 28 °C em Kelvin
Alpha = 0.01;
Cp = 500;
A = 0.0012;
m = 0.004;
U = 8;
Emissividade = 0.9;
Boltzmann = 5.67e-8;

%% Criação do gráfico
figure('Position', [100 100 800 600]);

%% Loop principal de aquisição e simulação
start_time = tic;
prev_time = toc(start_time);

for i = 1:loops
    % Espera até que se complete o intervalo de dt
    sleep_time = max(0.01, dt - (toc(start_time) - prev_time));
    pause(sleep_time);
    
    % Atualizar tempo
    t = toc(start_time);
    prev_time = t;
    tm(i) = t;
    
    % Leitura da temperatura real do TCLab
    T1(i) = T1C();
    
    % --- Modelo Não Linear ---
    if i == 1
        T1_ModeloNL(i) = T1(i); % Condição inicial
    else
        Taquecedor = T1_ModeloNL(i-1) + 273.15; % Converter para Kelvin
        dTdt = ((Alpha / (m * Cp)) * Q) + ((U * A) / (m * Cp)) * (Tambiente - Taquecedor) ...
              + ((Emissividade * Boltzmann * A) / (m * Cp)) * (Tambiente^4 - Taquecedor^4);
        T1_ModeloNL(i) = T1_ModeloNL(i-1) + dTdt * dt;
    end
    
    % --- Modelo Linear ---
    if i == 1
        T1_ModeloL(i) = T1(i); % Condição inicial
    else
        T1_ModeloL(i) = T1_ModeloL(i-1) + 0.00499 * exp(-tm(i) / 123) * Q;
    end
    
    % Atualizar gráfico
    clf;
    subplot(2,1,1);
    plot(tm(1:i), T1(1:i), 'r-', 'LineWidth', 2); hold on;
    plot(tm(1:i), T1_ModeloNL(1:i), 'b--', 'LineWidth', 2);
    plot(tm(1:i), T1_ModeloL(1:i), 'g--', 'LineWidth', 2);
    ylabel('Temperatura (°C)', 'FontSize', 14);
    legend('Temperatura Real', 'Modelo Não Linear', 'Modelo Linear', 'FontSize', 12);
    grid on;
    
    subplot(2,1,2);
    plot(tm(1:i), Q1(1:i), 'r-', 'LineWidth', 2);
    ylabel('Potência (%)', 'FontSize', 14);
    xlabel('Tempo (s)', 'FontSize', 14);
    legend('Potência (%)', 'FontSize', 12);
    grid on;
    
    drawnow;
end

%% Pós-processamento
erro_medio_NL = mean(abs(T1 - T1_ModeloNL));
erro_medio_L = mean(abs(T1 - T1_ModeloL));

fprintf('\nErro Médio Absoluto entre Temperatura Real e Modelo Não Linear: %.2f °C\n', erro_medio_NL);
fprintf('Erro Médio Absoluto entre Temperatura Real e Modelo Linear: %.2f °C\n', erro_medio_L);

% Salvar gráfico em PNG
saveas(gcf, 'Grafico_TCLab_Modelo_Linear.png');

% Salvar dados em TXT
dados = [tm, T1, T1_ModeloNL, T1_ModeloL, Q1];
header = {'Tempo_s', 'Temperatura_Real_C', 'Modelo_Nao_Linear_C', 'Modelo_Linear_C', 'Potencia_pct'};
T = array2table(dados, 'VariableNames', header);
writetable(T, 'Dados_simulacao_03.txt', 'Delimiter', '\t');

disp('Simulação concluída.');

%% Função de desligamento seguro
function desligarTCLab()
    try
        h1(0);
        led(0);
        disp('Dispositivos TCLab desligados com segurança.');
    catch
        disp('Erro ao tentar desligar os dispositivos.');
    end
end
